- Added sorting by due dates and course ids for tasks - Added unit tests for sorting - Ensured tasks tied to a course are deleted when the course is deleted

Firehead294 opened the pull request.

Firehead294 committed:
- Added sorting by due dates and course ids for tasks
- Added unit tests for sorting
- Ensured tasks tied to a course are deleted when the course is deleted

ChunPoon commented:
I spotted a potential problem in the source code for courses, in the test cases the start_date variable are supposed to be followed by end_date by in the source files it is the opposite (as in end_date before start_date)
![Swapped vars](https://github.com/user-attachments/assets/3474ad6d-cbf3-41d5-ac8d-f37eeb33bc60)

kevgbailey reviewed (changes requested):
Where in this pull request are tasks deleted when the course is deleted? I'm not seeing anything regarding that, other than that it looks fire

Firehead294 commented:
In main.py, in the delete_course() function, line 69: tasks = [task for task in tasks if task["course_id"] != course_id] removes any tasks associated with that course from the list.

Firehead294 committed:
Merge remote-tracking branch 'origin/CLIprototype' into rykergibbons

# Conflicts:
#	src/main.py

Firehead294 committed:
- Added sorting by due dates and course ids for tasks
- Added unit tests for sorting
- Ensured tasks tied to a course are deleted when the course is deleted

ChunPoon commented:
In task_tests.py, the test about creating a task with an invalid course_id of -1 assert the course_id will be -1 instead of throwing some error like ValueError, is this ok?
![image](https://github.com/user-attachments/assets/10022e69-10d0-449d-88b7-3dcd10fd2e7c)

Firehead294 commented:
I went ahead and resolved a conflict where in one of the sorting functions (don't remember which) to make sure that the list was being saved after being sorted. I also switched the orders of start_date and end_date in course.py for clarity. The -1 in the assert should be fine I think. How are we feeling so far?

ChunPoon commented:
Writing the test report and looking thru the test files to see any minor bugs if they are found. Other than that I am feeling good so far

extremepayne commented on src/main.py:
```diff
@@ -281,56 +283,42 @@ def parse_flags() -> None:
             -sc  Sort tasks by course ID
             """
             )
-        match flag:
```
Not sure why we got rid of this `match` statement. To me, it's more readable than the chain of `if`s.

Firehead294 replied:
It was changed to a string of if statements originally due to my IDE not having the latest version of Python. I resolved this recently, it has been converted to a match statement again for readability.

extremepayne commented on src/main.py:
```diff
@@ -281,56 +283,42 @@ def parse_flags() -> None:
+        # match flag:
+        if flag == "-cc":
+            create_course()
+        if flag == "-lc":
```
If we're going to stick with the chain of `if`s, this one and all subsequent should really be `elif`s.

extremepayne commented:
> In main.py, in the delete_course() function, line 69: tasks = [task for task in tasks if task["course_id"] != course_id] removes any tasks associated with that course from the list.

That's all well and good, but this PR doesn't touch `main.py` line 69. If it doesn't change it, it shouldn't be in the PR title.

Firehead294 committed:
- Added sorting by due dates and course ids for tasks
- Added unit tests for sorting

extremepayne commented:
lgtm :+1:
